### PR TITLE
Chore/add typedefs peerstore book template 2

### PR DESCRIPTION
**Pull request by Gozala**
This pull request does following:

- Adds aegir from https://github.com/ipfs/aegir/pull/689 so that new typescript@4.1.x is used
  - Which resolves issues with generic class extensions
- It addresses bunch of issues that type checker reported
- I merged #802 into which I think had some fixes that were missing here

@vasco-santos one other thing I want to call out here is that this change turns `StreamHandler` into non generic, that is because it uses `CircuitPB` encode / decode which in turn expects concrete structs. I am however not exactly sure I got all the details right, specifically:

1. According to `CircuitRelay` protobuf all fields are optional.
2. But than typedef in #802 was defining `CircuitRequest` with all fields (but `code`) mandatory.
4. Writes to `StreamHandler` seems to always contain `code & type`
5. Reads values from `StreamHandler` are assumed to have `type` `srcPeer`, `dstPeer`.

Which is what were reflected in the StreamHandler API reads return `CircuitRequest`s with all the mandatory fields and writes take `CircuitMessage`s with all the optional fields. However I am guessing that `CircuitMessage` in fact is more of a union of different message types where each corresponds to `Type` enum and has different subset of fields & probably each must have some subset of the the fields and that is why protobuf just marks them all optional.

If above hypothesis is correct it would make much more sense to represent each `message` type via actual type definition with it's fields and represent `CircuitMessage` as a type union. It would be good to sync up on this to fix those details up.

### CODE REVIEW

**Comment by vasco-santos** [comment]
Added a commit merging in the new updates of the base types branch to check this

File: src/upgrader.js [hunk 1]
```diff
@@ -231,6 +231,7 @@ class Upgrader {
             const { stream, protocol } = await mss.handle(Array.from(this.protocols.keys()))
             log('%s: incoming stream opened on %s', direction, protocol)
             if (this.metrics) this.metrics.trackStream({ stream, remotePeer, protocol })
+            // @ts-ignore - metadata seems to be required
```
Comment (vasco-santos) [inline src/upgrader.js:234]: Fixed on https://github.com/libp2p/js-libp2p-interfaces/pull/74

**Comment by vasco-santos** [comment]
This looks good to me! However, a new issue appeared as a result of this update...

```
File './node_modules/@types/node/events.d.ts' is not a module.
import { EventEmitter } from "node/events"
```

This only happens with the new ts. If you look into the dist on `libp2p/dist/src/connection-manager/index.d.ts:153:30` you will see the issue with vscode.

**Comment by vasco-santos** [comment]
> Which is what were reflected in the StreamHandler API reads return CircuitRequests with all the mandatory fields and writes take CircuitMessages with all the optional fields. However I am guessing that CircuitMessage in fact is more of a union of different message types where each corresponds to Type enum and has different subset of fields & probably each must have some subset of the the fields and that is why protobuf just marks them all optional.

> If above hypothesis is correct it would make much more sense to represent each message type via actual type definition with it's fields and represent CircuitMessage as a type union. It would be good to sync up on this to fix those details up.


Yes, this sounds good! I will add it to the follow up list to not block this anymore

**Comment by lidel** [comment]
Discussion continued in #830

**Pull request closed without merging**